tree/ntuple/v7/inc/ROOT/RColumn.hxx

```diff
@@ -176,51 +176,33 @@ public:
 
    void Read(const RClusterIndex &clusterIndex, void *to)
    {
-      if (!fReadPage.Contains(clusterIndex)) {
-         MapPage(clusterIndex);
-      }
-      const auto elemSize = fElement->GetSize();
-      void *from = static_cast<unsigned char *>(fReadPage.GetBuffer()) +
-                   (clusterIndex.GetIndex() - fReadPage.GetClusterRangeFirst()) * elemSize;
-      std::memcpy(to, from, elemSize);
+      Read(GetGlobalIndex(clusterIndex), to);
    }
 
-   void ReadV(const NTupleSize_t globalIndex, const ClusterSize_t::ValueType count, void *to)
+   void ReadV(NTupleSize_t globalIndex, ClusterSize_t::ValueType count, void *to)
    {
-      if (!fReadPage.Contains(globalIndex)) {
-         MapPage(globalIndex);
-      }
-      NTupleSize_t idxInPage = globalIndex - fReadPage.GetGlobalRangeFirst();
+      while (count > 0) {
+         if (!fReadPage.Contains(globalIndex)) {
+            MapPage(globalIndex);
+         }
+         NTupleSize_t idxInPage = globalIndex - fReadPage.GetGlobalRangeFirst();
 
-      const auto elemSize = fElement->GetSize();
-      const void *from = static_cast<unsigned char *>(fReadPage.GetBuffer()) + idxInPage * elemSize;
-      if (globalIndex + count <= fReadPage.GetGlobalRangeLast() + 1) {
-         std::memcpy(to, from, elemSize * count);
-      } else {
+         const auto elemSize = fElement->GetSize();
+         const void *from = static_cast<unsigned char *>(fReadPage.GetBuffer()) + idxInPage * elemSize;
          ClusterSize_t::ValueType nBatch = fReadPage.GetNElements() - idxInPage;
+         if (nBatch > count) {
+            nBatch = count;
+         }
          std::memcpy(to, from, elemSize * nBatch);
-         auto tail = static_cast<unsigned char *>(to) + nBatch * elemSize;
-         ReadV(globalIndex + nBatch, count - nBatch, tail);
+         globalIndex += nBatch;
+         count -= nBatch;
+         to = static_cast<unsigned char *>(to) + nBatch * elemSize;
       }
    }
 
    void ReadV(const RClusterIndex &clusterIndex, const ClusterSize_t::ValueType count, void *to)
    {
-      if (!fReadPage.Contains(clusterIndex)) {
-         MapPage(clusterIndex);
-      }
-      NTupleSize_t idxInPage = clusterIndex.GetIndex() - fReadPage.GetClusterRangeFirst();
-
-      const auto elemSize = fElement->GetSize();
-      const void *from = static_cast<unsigned char *>(fReadPage.GetBuffer()) + idxInPage * elemSize;
-      if (clusterIndex.GetIndex() + count <= fReadPage.GetClusterRangeLast() + 1) {
-         std::memcpy(to, from, elemSize * count);
-      } else {
-         ClusterSize_t::ValueType nBatch = fReadPage.GetNElements() - idxInPage;
-         std::memcpy(to, from, elemSize * nBatch);
-         auto tail = static_cast<unsigned char *>(to) + nBatch * elemSize;
-         ReadV(RClusterIndex(clusterIndex.GetClusterId(), clusterIndex.GetIndex() + nBatch), count - nBatch, tail);
-      }
+      ReadV(GetGlobalIndex(clusterIndex), count, to);
```

Comment:
There is a slight (but perhaps irrelevant) change in semantics here: If reduced to the global-index version, we can do a vector read across cluster boundaries, which we previously could not.

Reply:
Yes, good point. Not sure if relevant...